File: week-13/dahyen0o/징검다리-건너기.java
import java.util.*;

class Solution {
    public int solution(final int[] stones, final int K) {
        int answer = 0;
        int left = 0, right = 200_000_000;
        
        while(left <= right) {
            final int mid = (int) ((left + (long) right) / 2);
            
            // 징검다리를 건널 수 있는지 검사
            int k = 0;
            boolean canArrive = true;
            for(int i = 0; i < stones.length; i++) {
                final int stone = stones[i] - mid;
                if(stone <= 0) {
                    k += 1;
                    if(k >= K) {
                        canArrive = false;
                        break;
                    }
                } else {
                    k = 0;
                }
            }

            if(canArrive) {
                left = mid + 1;
                answer = Math.max(answer, mid);
            } else{
                right = mid - 1;
            }
        }
        
        return answer + 1;
    }
}

File: week-13/rimrim990/징검다리 건너기.py
def solution(stones, k):
  ans = 0
  start, end = 1, max(stones)
  while start <= end:
    mid = (start + end) // 2
    if check(stones, k, mid):
      ans = max(ans, mid)
      start = mid + 1

    else:
      end = mid - 1

  return ans

def check(stones, k, cnt):
  # 마지막 친구가 징검다리를 건너갈 수 있는지 확인
  remain = k
  for stone in stones:
    if stone <= cnt-1:
      remain -= 1
      if remain == 0: return False

    else:
      remain = k

  return True
